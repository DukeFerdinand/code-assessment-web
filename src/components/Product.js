import React from 'react'
import PropTypes from 'prop-types'
import ReactSVG from 'react-svg'

import images from '../assets/images'
import icons from '../assets/icons/index'

const Product = ({ price, inventory, title, quantity, inCart, removeProductFromCart }) => (
  <div className="product">
    {
      inCart ? (
        // Again as before, these images should be fed from a Database,
        // but that's outside the scope of this project
        <span className="product-image-wrapper">
          <img alt={title} className="product-image" src={images.thumbnails[`${title}Thumb`] || images.NotFound} />
        </span>
      ) : null
    }
    <div className="product-info-upper">
      <div className="product-title">
        {title}
      </div>
      <div className="product-price">
        &#36;{price}
      </div>
      { inCart ? <span onClick={() => removeProductFromCart()} className="remove-product-link">Remove</span> : null }
    </div>
    {
      !inCart ? (
        <div className={`product-availability ${inventory ? '' : 'out-of-stock'}`}>
          {inventory ? `${inventory} REMAINING` : 'OUT OF STOCK'}
        </div>
      ) : (
        <div className="product-quantity">
          <button
            className="minus-button"
            onClick={() => removeProductFromCart(1)}
            >
            <ReactSVG
              className="button-icon-wrapper"
              svgClassName="button-icon"
              src={icons.minus}
            />
          </button>
          <span className="product-quantity-input-wrapper">
            <input className="product-quantity-input" type="text" value={quantity}/>
          </span>
          <button className="plus-button">
            <ReactSVG
              className="button-icon-wrapper"
              svgClassName="button-icon"
              src={icons.plus}
            />
          </button>
        </div>
      )
    }
  </div>
)

Product.propTypes = {
  price: PropTypes.number,
  inventory: PropTypes.number,
  title: PropTypes.string,
  quantity: PropTypes.number,
  inCart: PropTypes.bool,
  removeProductFromCart: PropTypes.func
}

export default Product
