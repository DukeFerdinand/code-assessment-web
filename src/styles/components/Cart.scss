@import '../shared';

.display-cart {
  display: initial !important;
}

.cart-wrapper {
  background: rgba(20, 20, 20, .5);
  height: 100vh;
  width: 100vw;
  position: fixed;
  top: 0;
  left: 0;
  right: 0;
  margin-left: auto;
  margin-right: auto;

  display: none;
  .cart-container {
    z-index: 3;
    width: 100%;
    max-width: 500px;
    height: 100%;
    max-height: 500px;
    border-radius: 10px;
    box-sizing: border-box;
    padding: 50px;
    background: white;
    box-shadow: 0 0 5px 0 (black + 100);
    position: relative;
    left: 0;
    right: 0;
    margin: 50px auto;

    .cart-close-wrapper {
      position: absolute;
      top: 20px;
      right: 20px;
      cursor: pointer;

      .cart-close-icon {
        height: 25px;
        width: 25px;
      }
    }

    .cart-title {
      font-size: 1.53em;
    }

    .cart-content {
      height: 300px;
      .cart-empty-message {
        width: 200px;
        text-align: center;
        // max-width: 50px;
        position: absolute;
        left: 0;
        right: 0;
        margin: 0 auto;
        top: 40%;
        .cart-empty-icon-wrapper {
          margin: 0 auto;
          width: 75px;
          height: 75px;
          * {
            fill: $text-color-background;
          }

          .cart-empty-icon {
            width: 100%;
            height: 100%;
          }
        }
        .cart-empty-text {
          color: $text-color-background;
        }
      }
    }


    .divider {
      @extend %divider;
      border-top: 2px solid;
      border-color: black + 245 !important;
    }
  }
}