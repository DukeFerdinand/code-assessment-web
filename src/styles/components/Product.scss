@import '../shared';

.product-wrapper {
  width: 100%;
  height: auto;
  max-height: 275px;
  display: flex;
  margin: 30px auto;

  @media (max-width: $tablet) {
    max-height: 175px;
    margin: 20px auto;
  }
  .product-image {
    flex: .8;
    background: black + 200;
    height: 100%;
    max-height: 250px;
    max-width: 400px;
    border-radius: 10px 0 0 10px;

    @media (max-width: $tablet) {
      max-width: 272px;
      max-height: 175px;
      border-radius: 5px 0 0 5px;
    }
  }
  .product-info-wrapper {
    flex: 1;
    border-radius: 0 8px 8px 0;
    background: white;
    .product-info {
      box-sizing: border-box;
      height: 100%;
      padding: 30px 60px;
      display: flex;
      flex-direction: column;
      justify-content: space-between;

      @media (max-width: $tablet) {
        padding: 20px 20px 20px 40px;
      }

      .product-info-upper {
        display: flex;
        justify-content: space-between;

        .product-title {
          font-size: 2.5em;
          color: $text-color-main;

          @media (max-width: $tablet) {
            font-size: 1.5em;
          }
        }
        .product-price {
          font-size: 1.3em;
          position: relative;
          top: 10px;

          @media (max-width: $tablet) {
            font-size: 1.1em;
            top: 5px;
          }
        }
      }
      .product-availability {
        margin: 10px 5px;
        font-size: .90em;
        color: $text-color-secondary;
        &.out-of-stock {
          color: #e74c3c;
        }

        @media (max-width: $tablet) {
          font-size: .75em;
        }
      }
      .product-action {
        width: 130px;
        background: $color-primary;
        border-radius: 20px;
        color: white;
        cursor: pointer;
        font-size: .75em;
        padding: 10px 10px;

        @media (max-width: $tablet) {
          bottom: 0px;
          padding: 8px;
          width: 120px;
          font-size: .7em;
        }

        &:hover, &:focus {
          background: $button-primary-hover
        }

        &:disabled {
          cursor: initial;
          background: $color-primary-disabled;
        }
      }
    }
  }
}

.cart-content {
  .product {
    display: flex;
    // background: ;
    flex-wrap: wrap;
    margin-bottom: 25px;
    .product-image-wrapper {
      width: 150px;
      // height: 125px;
      .product-image {
        width: 100%;
      }
    }
    .product-info-upper {
      height: 100px;
      margin-left: 10px;
      position: relative;
      .product-title {
        font-size: 1.25em;
      }
      .product-price {
        margin-top: 5px;
      }
      .remove-product-link {
        // TODO: switch back to cursor
        cursor: not-allowed;
        flex: .01;
        color: red - 10;
        position: absolute;
        bottom: 10px;
      }
    }
    .product-quantity {
      display: flex;
      margin: 10px auto 0 auto;
      border-radius: 60px;

      %quantity-button {
        //TODO: Re-implement this once the cart buttons work
        // cursor: pointer;
        cursor: not-allowed;
        height: 40px;
        width: 80px;
        flex: 1;
        background: $button-secondary;
        // TODO: Re-implement once cart buttons work
        // &:hover {
        //   background: $button-primary;
        //   * {
        //     fill: white;
        //   }
        // }
      }
      .minus-button {
        @extend %quantity-button;
        border-radius: 30px 0 0 30px;
        .button-icon-wrapper {
          margin: 0 auto 5px auto;
          width: 15px;
          .button-icon {
            width: 15px;
          }
        }
      }
      .product-quantity-input-wrapper {
        background: $button-secondary;
        height: 40px;
        flex: 1;
        .product-quantity-input {
          border: none;
          font: $app-font;
          font-size: 1em;
          margin-top: 2px;
          height: 34px;
          width: 100%;
          text-align: center;
        }
      }

      .plus-button {
        @extend %quantity-button;
        border-radius: 0 30px 30px 0;

        .button-icon-wrapper {
          margin: 2px auto 0 auto;
          padding-right: 10px;
          width: 15px;
          .button-icon {
            width: 15px;
          }
        }
      }
    }
  }
}