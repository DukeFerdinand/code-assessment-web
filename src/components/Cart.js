import React from 'react'
import ReactSVG from 'react-svg'
import PropTypes from 'prop-types'
import Product from './Product'
import CartCheckout from './CartCheckout'

import CloseIcon from '../assets/close.svg'
import CartIcon from '../assets/cart.svg'

import '../styles/components/Cart.scss'

const Cart  = ({ products, total, onCloseCartClicked, onCheckoutClicked }) => {
  const hasProducts = products.length > 0
  const nodes = hasProducts ? (
    products.map(product =>
      <Product
        inCart={true}
        title={product.title}
        price={product.price}
        quantity={product.quantity}
        key={product.id}
      />
    )
  ) : (
    <div className="cart-empty-message">
      <span className="cart-empty-text">
        <ReactSVG
          src={CartIcon}
          className="cart-empty-icon-wrapper"
          svgClassName="cart-empty-icon"
        />
        Please add some products to your cart.
      </span>
    </div>
  )

  return (
    <div className="cart-container">
      <ReactSVG
        src={CloseIcon}
        onClick={onCloseCartClicked}
        className="cart-close-wrapper"
        svgClassName="cart-close-icon"
      />
      <h3 className="cart-title">Your Cart</h3>
      <hr className="divider" />
      <div className="cart-content-wrapper">
        <div className="cart-content">{nodes}</div>
        {
          hasProducts ?
            (
              <CartCheckout
                total={total}
                onCheckoutClicked={onCheckoutClicked}
              />
            ) : null
        }
      </div>
    </div>
  )
}

Cart.propTypes = {
  products: PropTypes.array,
  total: PropTypes.string,
  onCheckoutClicked: PropTypes.func,
  onCloseCartClicked: PropTypes.func
}

export default Cart
