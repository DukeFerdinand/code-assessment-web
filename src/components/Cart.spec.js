import React from 'react'
import { shallow } from 'enzyme'
import Cart from './Cart'
import Product from './Product'

const setup = (total, products = []) => {
  const actions = {
    onCheckoutClicked: jest.fn()
  }

  const component = shallow(
    <Cart products={products} total={total} {...actions} />
  )

  return {
    component: component,
    actions: actions,
    button: component.find('button') || null,
    products: component.find(Product),
    emptyText: component.find('span.cart-empty-text'),
    p: component.find('p.checkout-total')
  }
}

describe('Cart component', () => {

  describe('when not given a total or products', () => {
    it('should display add some products message', () => {
      const { emptyText } = setup()
      expect(emptyText.text()).toMatch(/<ReactSVG \/>Please add some products to your cart./)
    })

    it('should not display checkout button', () => {
      const { button } = setup()
      expect(button.length).toEqual(0)
    })
  });

  describe('when given product', () => {
    const product = [
      {
        id: 1,
        title: 'Product 1',
        price: 9.99,
        quantity: 1
      }
    ]
    it('should display total', () => {
      const { p } = setup('76', product)
      expect(p.text()).toMatch(/^Total: \$76/)
    })

    it('should render products', () => {
      const { products } = setup('9.99', product)
      const props = {
        title: product[0].title,
        price: product[0].price,
        quantity: product[0].quantity
      }

      expect(products.at(0).props()).toEqual(props)
    })

    it('should not disable button', () => {
      const { button } = setup('9.99', product)
      expect(button.prop('disabled')).toEqual('')
    })

    it('should call action on button click', () => {
      const { button, actions } = setup('9.99', product)
      button.simulate('click')
      expect(actions.onCheckoutClicked).toBeCalled()
    })
  })

  it('should match snapshot', () => {
    const product = [
      {
        id: 1,
        title: 'Product 1',
        price: 9.99,
        quantity: 1
      }
    ]
    const { component } = setup('9.99', product)
    expect(component).toMatchSnapshot()
  })
})
