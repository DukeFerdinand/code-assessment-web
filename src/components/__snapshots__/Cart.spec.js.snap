// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cart component should match snapshot 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <div
    className="cart-container"
  >
    <ReactSVG
      className="cart-close-wrapper"
      evalScripts="never"
      fallback={null}
      loading={null}
      onInjected={[Function]}
      renumerateIRIElements={true}
      src="close.svg"
      svgClassName="cart-close-icon"
      svgStyle={Object {}}
      wrapper="div"
    />
    <h3
      className="cart-title"
    >
      Your Cart
    </h3>
    <hr
      className="divider"
    />
    <div
      className="cart-content-wrapper"
    >
      <div
        className="cart-content"
      >
        <Product
          addToCart={[Function]}
          inCart={true}
          price={9.99}
          quantity={1}
          quantityInputHandler={[Function]}
          removeProductFromCart={[Function]}
          title="Product 1"
        />
      </div>
      <CartCheckout
        onCheckoutClicked={[MockFunction]}
        total="9.99"
      />
    </div>
  </div>,
  "nodes": Array [
    <div
      className="cart-container"
    >
      <ReactSVG
        className="cart-close-wrapper"
        evalScripts="never"
        fallback={null}
        loading={null}
        onInjected={[Function]}
        renumerateIRIElements={true}
        src="close.svg"
        svgClassName="cart-close-icon"
        svgStyle={Object {}}
        wrapper="div"
      />
      <h3
        className="cart-title"
      >
        Your Cart
      </h3>
      <hr
        className="divider"
      />
      <div
        className="cart-content-wrapper"
      >
        <div
          className="cart-content"
        >
          <Product
            addToCart={[Function]}
            inCart={true}
            price={9.99}
            quantity={1}
            quantityInputHandler={[Function]}
            removeProductFromCart={[Function]}
            title="Product 1"
          />
        </div>
        <CartCheckout
          onCheckoutClicked={[MockFunction]}
          total="9.99"
        />
      </div>
    </div>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 2,
      "_context": Object {},
      "_currentElement": <Cart
        onCheckoutClicked={[MockFunction]}
        products={
          Array [
            Object {
              "id": 1,
              "price": 9.99,
              "quantity": 1,
              "title": "Product 1",
            },
          ]
        }
        total="9.99"
      />,
      "_debugID": 9,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": StatelessComponent {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "props": Object {
          "onCheckoutClicked": [MockFunction],
          "products": Array [
            Object {
              "id": 1,
              "price": 9.99,
              "quantity": 1,
              "title": "Product 1",
            },
          ],
          "total": "9.99",
        },
        "refs": Object {},
        "state": null,
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 5,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <div
          className="cart-container"
        >
          <ReactSVG
            className="cart-close-wrapper"
            evalScripts="never"
            fallback={null}
            loading={null}
            onInjected={[Function]}
            renumerateIRIElements={true}
            src="close.svg"
            svgClassName="cart-close-icon"
            svgStyle={Object {}}
            wrapper="div"
          />
          <h3
            className="cart-title"
          >
            Your Cart
          </h3>
          <hr
            className="divider"
          />
          <div
            className="cart-content-wrapper"
          >
            <div
              className="cart-content"
            >
              <Product
                addToCart={[Function]}
                inCart={true}
                price={9.99}
                quantity={1}
                quantityInputHandler={[Function]}
                removeProductFromCart={[Function]}
                title="Product 1"
              />
            </div>
            <CartCheckout
              onCheckoutClicked={[MockFunction]}
              total="9.99"
            />
          </div>
        </div>,
        "_debugID": 10,
        "_renderedOutput": <div
          className="cart-container"
        >
          <ReactSVG
            className="cart-close-wrapper"
            evalScripts="never"
            fallback={null}
            loading={null}
            onInjected={[Function]}
            renumerateIRIElements={true}
            src="close.svg"
            svgClassName="cart-close-icon"
            svgStyle={Object {}}
            wrapper="div"
          />
          <h3
            className="cart-title"
          >
            Your Cart
          </h3>
          <hr
            className="divider"
          />
          <div
            className="cart-content-wrapper"
          >
            <div
              className="cart-content"
            >
              <Product
                addToCart={[Function]}
                inCart={true}
                price={9.99}
                quantity={1}
                quantityInputHandler={[Function]}
                removeProductFromCart={[Function]}
                title="Product 1"
              />
            </div>
            <CartCheckout
              onCheckoutClicked={[MockFunction]}
              total="9.99"
            />
          </div>
        </div>,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <Cart
    onCheckoutClicked={[MockFunction]}
    products={
      Array [
        Object {
          "id": 1,
          "price": 9.99,
          "quantity": 1,
          "title": "Product 1",
        },
      ]
    }
    total="9.99"
  />,
}
`;
